pydantic/function_model.py

```diff
@@ -0,0 +1,442 @@
+"""Provides FunctionModel, a model for validating function arguments."""
+
+from collections import OrderedDict
+from copy import deepcopy
+from inspect import Parameter, Signature, signature
+from types import NoneType
+from typing import Any, Callable, List, Optional, Tuple, Type, TypeVar, Union
+
+from pydantic_core import InitErrorDetails, PydanticCustomError, PydanticKnownError, ValidationError
+
+from pydantic import BaseModel, ConfigDict, TypeAdapter, create_model, errors
+
+
+class FunctionModel:
+    """A class for validating function arguments and calling the function upon the validated arguments."""
+
+    function: Callable
+    _model: Type[BaseModel]
+    _parsed: Optional[BaseModel]
+    _return: TypeAdapter
+    _parameters: OrderedDict
+    _keyword_total: int
+    _positional_total: int
+
+    T = TypeVar('T')
```

Comment:
Probably best to have this `TypeVar` outside of the class.

Reply:
Done.